File: android/app/src/main/kotlin/org/fgsec/halogen/components/Artwork.kt
package org.fgsec.halogen.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import coil3.compose.AsyncImage
import coil3.compose.LocalPlatformContext

/// Square rounded artwork tile with a placeholder. URLs point at the server's
/// art cache; ArtLoader fetches them with the API bearer token.
@Composable
fun Artwork(url: String?, size: Dp) {
    // iOS @ScaledMetric parity: the tile grows with the UI-size pref so a
    // bigger UI doesn't pair big labels with shrunken-looking art.
    val scaled = size * LocalDensity.current.fontScale
    Box(
        Modifier
            .size(scaled)
            .clip(RoundedCornerShape(scaled / 7))
            .background(halogenExtras.placeholderFill),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            halogenIcon("waveform"),
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        if (url != null) {
            AsyncImage(
                model = url,
                contentDescription = null,
                imageLoader = ArtLoader.loader(LocalPlatformContext.current),
                contentScale = ContentScale.Crop,
                modifier = Modifier.matchParentSize(),
            )
        }
    }
}

File: ios/Halogen/Components/Artwork.swift
import SwiftUI

/// Square rounded artwork tile with a placeholder. URLs point at the server's
/// art cache; ArtLoader fetches them with the API bearer token.
struct Artwork: View {
    let url: URL?
    /// Scales with the UI-size pref (Dynamic Type): callers pass the base
    /// (100%) point size and the tile grows with the text around it, so a
    /// bigger UI doesn't pair big labels with shrunken-looking art.
    @ScaledMetric private var size: CGFloat

    @State private var image: UIImage?

    init(url: URL?, size: CGFloat) {
        self.url = url
        _size = ScaledMetric(wrappedValue: size, relativeTo: .body)
    }

    var body: some View {
        ZStack {
            if let image {
                Image(uiImage: image).resizable().scaledToFill()
            } else {
                Rectangle().fill(.quaternary)
                Image(systemName: "waveform")
                    .foregroundStyle(.secondary)
            }
        }
        .frame(width: size, height: size)
        .clipShape(RoundedRectangle(cornerRadius: size / 7))
        .task(id: url) {
            image = nil
            guard let url else { return }
            let loaded = await ArtLoader.shared.image(for: url)
            guard !Task.isCancelled else { return }
            image = loaded
        }
    }
}
